src/irmin-pack/unix/control_file.ml

```diff
@@ -299,35 +299,38 @@ module Make (Serde : Serde.S) (Io : Io.S) = struct
   module Io = Io
 
   type payload = Serde.payload
-  type t = { io : Io.t; mutable payload : payload }
+  type t = { mutable io : Io.t; mutable payload : payload; path : string }
 
   let write io payload =
     let s = Serde.to_bin_string payload in
-    (* The data must fit inside a single page for atomic updates of the file.
-       This is only true for some file systems. This system will have to be
-       reworked for [V4]. *)
-    assert (String.length s <= Io.page_size);
-
     Io.write_string io ~off:Int63.zero s
 
+  let set_payload t payload =
+    let open Result_syntax in
+    let* () = Io.close t.io in
+    let path_tmp = t.path ^ ".tmp" in
+    let* io_tmp = Io.create ~path:path_tmp ~overwrite:true in
+    t.io <- io_tmp;
+    let* () = write io_tmp payload in
+    let+ () = Io.move_file ~src:path_tmp ~dst:t.path in
+    t.payload <- payload
+
   let read io =
     let open Result_syntax in
     let* string = Io.read_all_to_string io in
-    (* Since the control file is expected to fit in a page,
-       [read_all_to_string] should be atomic for most filesystems. *)
     Serde.of_bin_string string
 
   let create_rw ~path ~overwrite (payload : payload) =
     let open Result_syntax in
     let* io = Io.create ~path ~overwrite in
     let+ () = write io payload in
-    { io; payload }
+    { io; payload; path }
 
   let open_ ~path ~readonly =
     let open Result_syntax in
     let* io = Io.open_ ~path ~readonly in
     let+ payload = read io in
-    { io; payload }
+    { io; payload; path }
```

Comment:
Should we perhaps pass the `path ^ ".tmp"` as a `~tmp_path` on create/open? (but not as a concatenation, rather as a call to a function from `Layout` with proper categorization of the temp control file as "to be removed" within `File_manager.cleanup`?)

Reply:
Done